pkgs/ffigen/lib/src/header_parser/sub_parsers/api_availability.dart

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:ffi';
+
+import 'package:ffi/ffi.dart';
+
+import '../../config_provider/config_types.dart';
+import '../clang_bindings/clang_bindings.dart' as clang_types;
+import '../data.dart';
+import '../utils.dart';
+
+bool isApiAvailable(clang_types.CXCursor cursor) {
+  final api = ApiAvailability.fromCursor(cursor);
+  return api.isAvailable(config.objCMinTargetVersion);
+}
+
+class ApiAvailability {
+  final bool alwaysDeprecated;
+  final bool alwaysUnavailable;
+  PlatformAvailability? ios;
+  PlatformAvailability? macos;
+
+  ApiAvailability({
+    this.alwaysDeprecated = false,
+    this.alwaysUnavailable = false,
+    this.ios,
+    this.macos,
+  });
+
+  static ApiAvailability fromCursor(clang_types.CXCursor cursor) {
+    final platformsLength = clang.clang_getCursorPlatformAvailability(
+        cursor, nullptr, nullptr, nullptr, nullptr, nullptr, 0);
+
+    final alwaysDeprecated = calloc<Int>();
+    final alwaysUnavailable = calloc<Int>();
+    final platforms =
+        calloc<clang_types.CXPlatformAvailability>(platformsLength);
+
+    clang.clang_getCursorPlatformAvailability(cursor, alwaysDeprecated, nullptr,
+        alwaysUnavailable, nullptr, platforms, platformsLength);
+
+    PlatformAvailability? ios;
+    PlatformAvailability? macos;
+
+    for (var i = 0; i < platformsLength; ++i) {
+      final platform = platforms[i];
+      final platformAvailability = PlatformAvailability(
+        introduced: platform.Introduced.triple,
+        deprecated: platform.Deprecated.triple,
+        obsoleted: platform.Obsoleted.triple,
+        unavailable: platform.Unavailable != 0,
+      );
+      switch (platform.Platform.string()) {
+        case 'ios':
+          ios = platformAvailability;
+          break;
+        case 'macos':
+          macos = platformAvailability;
+          break;
+      }
+    }
+
+    final api = ApiAvailability(
+      alwaysDeprecated: alwaysDeprecated.value != 0,
+      alwaysUnavailable: alwaysUnavailable.value != 0,
+      ios: ios,
+      macos: macos,
+    );
+
+    for (var i = 0; i < platformsLength; ++i) {
+      clang.clang_disposeCXPlatformAvailability(platforms + i);
+    }
+    calloc.free(alwaysDeprecated);
+    calloc.free(alwaysUnavailable);
+    calloc.free(platforms);
+
+    return api;
+  }
+
+  bool isAvailable(ObjCTargetVersion minVers) {
+    // If no minimum versions are specified, everything is available.
+    if (minVers.ios == null && minVers.macos == null) {
+      return true;
+    }
+
+    if (alwaysDeprecated || alwaysUnavailable) {
+      return false;
+    }
+
+    for (final (plat, minVer) in [(ios, minVers.ios), (macos, minVers.macos)]) {
+      // If the user hasn't specified a minimum version for this platform, defer
```

Comment:
If users don't use the `@Native` but the `DynamicLibrary` version. Should that constructor of the bindings maybe take a `currentVersions` param that either contains the OS version of iOS or MacOS (or maybe better contain a map, so that we also know the OS properly). Then we can generate asserts / throws on APIs that are only available in a part of the range or on one OS.

We need to still figure out where to get this version from, but putting it on the dylib constructor pushes the responsibility to the user.

Maybe you can take a look at how `package:device_info_plus` gets the current MacOS and iOS version and see if we can do that within the generated code so that we don't have to expose it in the public API.